### ISSUE TITLE
RangeError in nodejs using arduinoWebsockets

### ISSUE THREAD

**Issue by tmarkson** [Links2004/arduinoWebSockets#533]
`RangeError: Invalid WebSocket frame: RSV2 and RSV3 must be clear`

Does this error look familiar coming from nodejs server? 

The client is a ESP8266 with the `setDebugOutput` enabled and there's an event for `onStationModeDisconnected`. Neither of those fire before the server fails. The client tries to run `sendBIN()` a few times unsuccessful before the server errors and crashes.

**Comment by tmarkson** [comment]
Here's the Wifi debug output.  I'm not sure about the opcodes and other shortcodes. Something I notice is about the two message frames that were sent before the frame that timed out. They took much longer to send than previous frames -- 122221us and 62498us compared to 8580us or even 27767us. 

> b'[WS][0][handleWebsocketWaitFor] size: 2 cWsRXsize: 0\n'
> b'[readCb] n: 2 t: 1199373\n'
> b'[WS][0][handleWebsocketWaitFor][readCb] size: 2 ok: 1\n'
> b'[WS][0][handleWebsocket] ------- read massage frame -------\n'
> b'[WS][0][handleWebsocket] fin: 1 rsv1: 0 rsv2: 0 rsv3 0  opCode: 1\n'
> b'[WS][0][handleWebsocket] mask: 0 payloadLen: 5\n'
> b'[readCb] n: 5 t: 1199375\n'
> b'[WS][0][handleWebsocket] text: ready\n'
> b'[WS][0][sendFrame] ------- send message frame -------\n'
> b'[WS][0][sendFrame] fin: 1 opCode: 2 mask: 1 length: 2000 headerToPayload: 0\n'
> b'[write] n: 8 t: 1199651\n'
> b'[write] n: 2000 t: 1199652\n'
> b'[WS][0][sendFrame] sending Frame Done (8580us).\n'
> 
> 
> b'[WS][0][handleWebsocketWaitFor] size: 2 cWsRXsize: 0\n'
> b'[readCb] n: 2 t: 1199664\n'
> b'[WS][0][handleWebsocketWaitFor][readCb] size: 2 ok: 1\n'
> b'[WS][0][handleWebsocket] ------- read massage frame -------\n'
> b'[WS][0][handleWebsocket] fin: 1 rsv1: 0 rsv2: 0 rsv3 0  opCode: 1\n'
> b'[WS][0][handleWebsocket] mask: 0 payloadLen: 5\n'
> b'[readCb] n: 5 t: 1199666\n'
> b'[WS][0][handleWebsocket] text: ready\n'
> b'[WS][0][sendFrame] ------- send message frame -------\n'
> b'[WS][0][sendFrame] fin: 1 opCode: 2 mask: 1 length: 2000 headerToPayload: 0\n'
> b'[write] n: 8 t: 1199966\n'
> b'[write] n: 2000 t: 1199967\n'
> b'[WS][0][sendFrame] sending Frame Done (27767us).\n'
> 
> 
> b'[WS][0][handleWebsocketWaitFor] size: 2 cWsRXsize: 0\n'
> b'[readCb] n: 2 t: 1200000\n'
> b'[WS][0][handleWebsocketWaitFor][readCb] size: 2 ok: 1\n'
> b'[WS][0][handleWebsocket] ------- read massage frame -------\n'
> b'[WS][0][handleWebsocket] fin: 1 rsv1: 0 rsv2: 0 rsv3 0  opCode: 1\n'
> b'[WS][0][handleWebsocket] mask: 0 payloadLen: 5\n'
> b'[readCb] n: 5 t: 1200002\n'
> b'[WS][0][handleWebsocket] text: ready\n'
> b'[WS][0][sendFrame] ------- send message frame -------\n'
> b'[WS][0][sendFrame] fin: 1 opCode: 2 mask: 1 length: 2000 headerToPayload: 0\n'
> b'[write] n: 8 t: 1200264\n'
> b'[write] n: 2000 t: 1200265\n'
> b'[WS][0][sendFrame] sending Frame Done (9076us).\n'
> 
> 
> b'[WS][0][handleWebsocketWaitFor] size: 2 cWsRXsize: 0\n'
> b'[readCb] n: 2 t: 1200283\n'
> b'[WS][0][handleWebsocketWaitFor][readCb] size: 2 ok: 1\n'
> b'[WS][0][handleWebsocket] ------- read massage frame -------\n'
> b'[WS][0][handleWebsocket] fin: 1 rsv1: 0 rsv2: 0 rsv3 0  opCode: 1\n'
> b'[WS][0][handleWebsocket] mask: 0 payloadLen: 5\n'
> b'[readCb] n: 5 t: 1200285\n'
> b'[WS][0][handleWebsocket] text: ready\n'
> b'[WS][0][sendFrame] ------- send message frame -------\n'
> b'[WS][0][sendFrame] fin: 1 opCode: 2 mask: 1 length: 2000 headerToPayload: 0\n'
> b'[write] n: 8 t: 1200564\n'
> b'[write] n: 2000 t: 1200565\n'
> b'[WS][0][sendFrame] sending Frame Done (6812us).\n'
> 
> 
> b'[WS][0][handleWebsocketWaitFor] size: 2 cWsRXsize: 0\n'
> b'[readCb] n: 2 t: 1200580\n'
> b'[WS][0][handleWebsocketWaitFor][readCb] size: 2 ok: 1\n'
> b'[WS][0][handleWebsocket] ------- read massage frame -------\n'
> b'[WS][0][handleWebsocket] fin: 1 rsv1: 0 rsv2: 0 rsv3 0  opCode: 1\n'
> b'[WS][0][handleWebsocket] mask: 0 payloadLen: 5\n'
> b'[readCb] n: 5 t: 1200581\n'
> b'[WS][0][handleWebsocket] text: ready\n'
> b'[WS][0][sendFrame] ------- send message frame -------\n'
> b'[WS][0][sendFrame] fin: 1 opCode: 2 mask: 1 length: 2000 headerToPayload: 0\n'
> b'[write] n: 8 t: 1200858\n'
> b'[write] n: 2000 t: 1200859\n'
> b'[WS][0][sendFrame] sending Frame Done (122221us).\n'
> 
> 
> b'[WS][0][handleWebsocketWaitFor] size: 2 cWsRXsize: 0\n'
> b'[readCb] n: 2 t: 1200990\n'
> b'[WS][0][handleWebsocketWaitFor][readCb] size: 2 ok: 1\n'
> b'[WS][0][handleWebsocket] ------- read massage frame -------\n'
> b'[WS][0][handleWebsocket] fin: 1 rsv1: 0 rsv2: 0 rsv3 0  opCode: 1\n'
> b'[WS][0][handleWebsocket] mask: 0 payloadLen: 5\n'
> b'[readCb] n: 5 t: 1200992\n'
> b'[WS][0][handleWebsocket] text: ready\n'
> b'[WS][0][sendFrame] ------- send message frame -------\n'
> b'[WS][0][sendFrame] fin: 1 opCode: 2 mask: 1 length: 2000 headerToPayload: 0\n'
> b'[write] n: 8 t: 1201158\n'
> b'[write] n: 2000 t: 1201159\n'
> b'[WS][0][sendFrame] sending Frame Done (62498us).\n'
> 
> 
> b'[WS][0][handleWebsocketWaitFor] size: 2 cWsRXsize: 0\n'
> b'[readCb] n: 2 t: 1201250\n'
> b'[WS][0][handleWebsocketWaitFor][readCb] size: 2 ok: 1\n'
> b'[WS][0][handleWebsocket] ------- read massage frame -------\n'
> b'[WS][0][handleWebsocket] fin: 1 rsv1: 0 rsv2: 0 rsv3 0  opCode: 1\n'
> b'[WS][0][handleWebsocket] mask: 0 payloadLen: 5\n'
> b'[readCb] n: 5 t: 1201252\n'
> b'[WS][0][handleWebsocket] text: ready\n'
> b'[WS][0][sendFrame] ------- send message frame -------\n'
> b'[WS][0][sendFrame] fin: 1 opCode: 2 mask: 1 length: 2000 headerToPayload: 0\n'
> b'[write] n: 8 t: 1201474\n'
> b'[write] n: 2000 t: 1201475\n'
> b'[write] write TIMEOUT! 5228\n'
> b'[WS][0][sendFrame] sending Frame Done (10695012us).\n'
> 
> Error with websocket transfer, resend.
> b'[WS][0][sendFrame] ------- send message frame -------\n'
> b'[WS][0][sendFrame] fin: 1 opCode: 2 mask: 1 length: 2000 headerToPayload: 0\n'
> b'[write] n: 8 t: 1212237\n'
> b'[write] n: 2000 t: 1212328\n'
> b'[WS][0][sendFrame] sending Frame Done (882450us).\n'
> 
> 
> b'[WS][0][sendFrame] ------- send message frame -------\n'
> b'[WS][0][sendFrame] fin: 1 opCode: 2 mask: 1 length: 2000 headerToPayload: 0\n'
> b'[write] n: 8 t: 1213170\n'
> b'[write] n: 2000 t: 1213170\n'
> b'[write] not connected!\n'
> b'[WS][0][sendFrame] sending Frame Done (23094us).\n'
> 
> Error with websocket transfer, resend.
> b'[WS-Client] connection lost.\n'
> b'[WS-Client] client disconnected.\n'
> b'[WSc] Disconnected!\n'
> b'[WS-Client] connect ws...\n'
> b'[hostByName] Host: 10.0.0.153 is a IP!\n'
> b'[WS-Client] connection to 10.0.0.153:5678 Failed\n'
> b'[WS-Client] client disconnected.\n'
> b'[WSc] Disconnected!\n'`

**Comment by Links2004** [comment]
the write timeout is the main problem here.
since not all data is transferred the parsing on the server ends up at the wrong point and finds the non zero RSV2 and RSV3 bits, which results in the error.

possible reasons:
- bad WiFi connection 
- bad or to less uplink  speed to the server
- to many messages at once (ESP TCP stack overload)

you may can improve it be reducing the data size you send (if possible) to less then the MTU,
with 2000 bytes the TCP stack is forced to split the message in multible TCP packets.

**Comment by tmarkson** [comment]
Thanks for the thoughts Markus
The AP is a home router with a few quiet devices on it. Speedtest shows 12Mbps up. And I would expect to see a disconnect message on Serial debug if the Wifi dropped. 

That's a good suggestion to lower the message size, I will test that now.  
I assume you're playing it safe at 1400B to be less than Ethernet's 1518B standard packet size?  I don't see a standard MTU value for Websockets after a search so I guess it's limited by the lower layers. 

It's strange that this started happening recently. It never happened in earlier testing; I tried message sizes of 256B to 10KB.

**Comment by Links2004** [comment]
it 1400Byte payload comes from this calculation:
- 1400B payload 
- 14B WS header 
- 40B TCP (TCP header has up to 40b but can be 20b)
- 20B IP
- 18B ETH
**1492B total**

which is what a typical DSL / PPPoE connection can handle

side note:
that the ESP TCP stack is limited to 1500 MTU max
(or 576 MTU based on what LwIPVariant you select in the IDE)

**Comment by tmarkson** [comment]
Would you link to where the MTU sizes are specified for the IP variants?  I'm using v2 Lower memory.

**Comment by Links2004** [comment]
the code defines the MSS which indirect defines the MTU.
the low memory is at MSS=536
https://github.com/esp8266/Arduino/blob/1127a090adc1a2469068f57d78db79f0c2d636df/boards.txt#L388-L417

**Comment by tmarkson** [comment]
Well that might be the issue.  Maybe I should use a size like 430 so the total frame size is less than 536B
. 
I've run a bunch of tests today and suddenly the error is not happening.... at all. Not once since I last posted here. I've restarted the server many times.  I did test it with 1400B payload and while it's slower there's no problem using that size for safety. 

Anyway I will keep running tests and reopen if it comes up again. Your info was very helpful Markus! Thank you
